### ISSUE TITLE
multi head training investigation

### ISSUE THREAD

**Issue by Talor-A** [Talor-A/forge#14]
<img width="1312" height="860" alt="Image" src="https://github.com/user-attachments/assets/b5e682b4-5bf3-409e-b8d6-0d5ba6854516" />

  a problem we're running into with the current experiment is that some heads converge drastically
  faster than others because the volume of training data per head is different by orders of
  magnitude, among other factors (collect is still in progress for today's run, but we can
  ballpark from yesterday's run).

  what's the correct solution to this?

---

Got it — stepping back from code. Here's the full picture.

## ELI5: what's actually going wrong

You're training one network with several "heads" — think one brain with several specialized output muscles. They share a single body (the **encoder**), and each muscle (head) learns one type of decision: priority actions, attacks, blocks, targets, mulligans, etc.

The problem: your training data is wildly lopsided. Priority has ~720k examples; target and block have ~21k each. That's a **34× imbalance**. And the way training currently loops, this imbalance causes *two separate problems* that are easy to lump together but actually have different causes and different fixes.

### Problem 1 — Small heads don't get enough practice

The trainer loops "round-robin": each cycle, it pulls one batch from every head that still has data left. With batch size 256:

- Priority: 720k ÷ 256 ≈ **2,810 batches per epoch**
- Target/block: 21k ÷ 256 ≈ **82 batches per epoch**

So the small heads take ~34× fewer training steps. Worse, your learning-rate schedule (cosine annealing) winds the LR down based on the *epoch number*, not on how many steps a head has actually taken. So the small heads get their learning rate cranked toward zero before they've had enough reps to learn. Block "underfits" for exactly this reason — it's not that block is hard, it's that block barely got to train.

Think of a gym class where everyone does the same number of *sessions*, but in each session the big group does 34 sets and the small group does 1 — then everyone's told "okay, cooldown" at the same time. The small group never got going.

### Problem 2 — The shared body gets hijacked by the biggest head

This is the subtler and arguably more important one. The encoder (the shared body) learns from *all* the heads' gradients — every head tugs the shared representation toward what's useful for it.

But here's the kicker in the current loop: once a head runs out of data, it's dropped (`active_heads -= finished`). The small heads exhaust after ~82 cycles. Priority keeps going **solo** for the remaining ~2,730 cycles. So roughly **97% of the updates to the shared body are shaped by priority alone.**

The result: the encoder becomes a *priority specialist*. The small heads are then trying to read decisions off an internal representation that was sculpted almost entirely for someone else's job. Even a well-trained target head is handicapped if the body it sits on only "thinks in priority."

Analogy: five specialists share one notebook of observations about the game. Four of them jot a few notes, then leave. The fifth stays all night and rewrites 97% of the notebook in their own shorthand. The other four come back to a notebook that's nearly useless to them.

**Why separating these two matters:** you could "fix" problem 1 (give small heads more steps) and still have problem 2 (encoder captured by priority), or vice versa. A good solution addresses both, but you should know which lever moves which.

---

### Option A — Temperature-scaled sampling (the industry-standard fix)

**Idea:** Stop letting "epoch = drain every head." Instead, define an epoch as a fixed number of steps *K*. At each step, randomly pick which head to train, with probability:

```
P(head) ∝ (N_head) ^ (1/T)
```

- **T = 1** → probability proportional to data size = today's behavior (priority dominates).
- **T = ∞** → uniform = every head equally likely regardless of data.
- **T ≈ 2–5** → the sweet spot most large multi-task systems land on.

T is a single dial between "let the data speak" and "treat every task equally." Small heads now get sampled far more often relative to their size, **and** because they're interleaved across the *whole* epoch (not just the first 3%), the encoder sees a balanced gradient mix throughout — so it fixes **both** problems with one mechanism.

- **Tradeoff:** Priority's own accuracy may dip slightly, because it no longer monopolizes the encoder. Whether that dip matters is an empirical question (does it hurt head-to-head win rate?) — which is exactly why sweeping T is smart rather than guessing.
- **Cost:** Cheap. It's a sampling change, not an architecture change.
- **Who uses it:** mT5, NLLB, and basically every large multilingual/multi-task model with order-of-magnitude data imbalance. This is *the* default answer to your exact question.

**Practical wrinkle worth flagging for the refactor:** when T pushes the small heads' target step-count *above* their dataset size (very likely at T≥2), you need to sample **with replacement / reshuffle across epochs** (oversample), not just take subsets. Your existing `EpochRerollSubsetSampler` only *down*-samples (it caps `n_keep` at `n_total`). So the refactor needs an oversampling-capable sampler, not just a generalization of the current one.

### Option B — Loss weighting

**Idea:** Train as today, but multiply each head's loss by a weight (e.g., upweight small heads) so they tug the encoder harder per step.

- **Tradeoff:** Fixes problem 2 (encoder balance) partially, but **not** problem 1 — small heads still take few steps; you've just made each step louder, which can make them noisier/unstable. Picking weights by hand is fiddly.
- **Verdict:** Usually inferior to sampling for *data-count* imbalance. Loss weighting is better suited to imbalance in *importance* (this task matters more) rather than imbalance in *volume*.

### Option C — Staged training (pretrain → freeze → fine-tune)

**Idea:** First train the encoder hard on the big head (priority). Then **freeze the encoder** and train each small head on top of the frozen body. Your code already has the `--freeze-encoder` flag for the second phase.

- **Tradeoff:** Explicitly accepts "the body is a priority specialist" and makes the small heads adapt to it. Two phases = more orchestration, more checkpoints to manage. Best when you *know* priority should dominate (e.g., it's the most frequent and most outcome-determining decision) and you just want the small heads to do their best on top.
- **When to reach for it:** If after the sampling fix the per-head best-epochs diverge wildly, or priority's win-rate contribution craters — that divergence is the signal that one shared schedule can't serve everyone, and staging is the principled escape hatch.

### Option D — Gradient-surgery methods (GradNorm, PCGrad, uncertainty weighting)

**Idea:** Algorithmically detect when heads' gradients conflict or when one dominates, and rescale/project to balance them automatically.

- **Tradeoff:** Real techniques, genuinely used — but **overkill here** for one-shot tuning. They add complexity and compute, and temperature sampling already captures most of the available win for data-count imbalance. Reach for these only if a balanced-sampling run shows heads actively *fighting* each other (negative transfer), not merely under-trained.

---

## The constraint you can't engineer away

This is the most important conceptual point, and it shapes the refactor. **Because the encoder is shared, there is no setting where every head independently picks its own best epoch.** The encoder can be tuned for *one* objective — priority-dominated (T=1), balanced (T≈2–5), or staged-then-frozen — but all heads then live or die on that one shared schedule. Your per-epoch-checkpoint scheme ("pick the best single coherent epoch post-hoc") only makes sense under a *single* shared objective; it cannot give you a per-head Pareto-best out of one shared encoder. If you ever truly need each head at its own optimum, that's a different architecture (separate encoders, or frozen-encoder staging), not a tuning knob.

---

## Industry general best practices (the short version)

1. **For data-volume imbalance, fix the sampling, not the loss.** Temperature/proportional sampling is the first-line tool; loss weighting is for *importance* imbalance.
2. **Decouple "epoch" from "data exhaustion."** Define an epoch as a fixed step budget and sample into it. This is what makes imbalanced multi-task training stable and comparable across runs.
3. **Sample with replacement for the minority tasks** so small heads aren't starved, and reshuffle each epoch so you still cover their full (small) dataset over time.
4. **Tie the LR schedule to steps, not wall-clock epochs**, when step counts differ across components — otherwise you anneal under-trained heads prematurely. (This is a real second-order issue in your current setup worth fixing in the refactor.)
5. **Make the imbalance knob (T) a swept hyperparameter, not a guessed constant.** T=1 should reproduce the naive baseline so your sweep is honestly anchored.
6. **Escalate in order of complexity:** sampling → loss weighting → staged/frozen → gradient surgery. Don't jump to (4) when (1) is untried.
7. **Shared encoder = shared destiny.** Decide deliberately whether the backbone is a generalist or a specialist; don't let the data-size accident decide it for you (which is what's happening today).

---

For the refactor, the cleanest target is: a `MultiTaskSampler` that takes per-head dataset sizes + a temperature `T` + a step budget `K`, and yields a `(head, batch)` stream — with oversampling for minority heads and step-based LR scheduling. That single component replaces the round-robin + per-head reroll-subsample logic and makes T=1 exactly reproduce today's behavior. Happy to sketch that interface (no implementation) whenever you pick the refactor up.
